src/main/scala/com/astrolabsoftware/grafink/services/IDManagerSparkService.scala

```diff
@@ -0,0 +1,143 @@
+     * processed
+     * @return
+     */
+    def readAll(schema: StructType, tableName: String): ZIO[Logging, Throwable, DataFrame]
+
+    /**
+     * Processes the input data for this job to generate and add new ids, then saves this data
+     * @param df
+     * @param tableName
+     * @return
+     */
+    def process(df: DataFrame, tableName: String): ZIO[Logging, Throwable, VertexData]
+    def fetchID(df: DataFrame): RIO[Logging, IDType]
+  }
+
+  val live: URLayer[Logging with SparkEnv with Has[IDManagerConfig], IDManagerSparkService] =
+    ZLayer.fromEffect(
+      for {
+        spark  <- ZIO.access[SparkEnv](_.get.sparkEnv)
+        config <- Config.idManagerConfig
+      } yield new IDManagerSparkServiceLive(spark, config)
+    )
+
+  def readAll(schema: StructType, tableName: String): RIO[IDManagerSparkService with Logging, DataFrame] =
+    RIO.accessM(_.get.readAll(schema, tableName))
+
+  def fetchID(df: DataFrame): RIO[IDManagerSparkService with Logging, IDType] =
+    RIO.accessM(_.get.fetchID(df))
+
+  def process(df: DataFrame, tableName: String): ZIO[IDManagerSparkService with Logging, Throwable, VertexData] =
+    RIO.accessM(_.get.process(df, tableName))
+}
+
+final class IDManagerSparkServiceLive(spark: SparkSession, config: IDManagerConfig)
+    extends IDManagerSparkService.Service {
+
+  def addId(df: DataFrame, lastMax: IDType): DataFrame =
+    SparkExtensions.zipWithIndex(df, lastMax + 1)
+
+  override def readAll(schema: StructType, tableName: String): ZIO[Logging, Throwable, DataFrame] =
+    ZIO.effect(spark.read.parquet(s"${config.spark.dataPath}/$tableName")).catchSome {
+      // Catch case where there is no data to read, this means this is being run on a new setup
+      case e: org.apache.spark.sql.AnalysisException if e.message.contains("Unable to infer schema for Parquet") =>
+        for {
+          _ <- log.warn(s"No data found at ${config.spark.dataPath}/$tableName, returning empty dataframe")
+        } yield spark.createDataFrame(
+          spark.sparkContext.emptyRDD[Row],
+          StructType(StructField("id", LongType, false) +: schema.fields)
+        )
+    }
+
+  override def process(df: DataFrame, tableName: String): ZIO[Logging, Throwable, VertexData] =
+    for {
+      // All the idmanager data so far ingested
+      idManagerDf <- readAll(df.schema, tableName)
+      // Get the last max id used
+      lastMax <- fetchID(idManagerDf)
+      // Generate new ids for the `jobTime` data
+      dfWithId <- processData(lastMax, df, tableName)
+    } yield VertexData(loaded = idManagerDf, current = dfWithId)
+
+  def processData(id: IDType, df: DataFrame, tableName: String): ZIO[Logging, Throwable, DataFrame] = {
+
+    val mode = SaveMode.Append
+
+    val dfWithId = addId(df, id)
+
+    for {
+      // Cache this df
+      dfWithIdCached <- ZIO.effect(dfWithId.cache)
+      // Write intermediate data
+      // TODO: Repartition to generate desired number of output files
```

Comment:
If i understand correctly, the default partitioning is the same as the input parquet files?

Reply:
Yes, so when we write we can end up with those many files. Ok let me ignore this for now